src/protocols/bdx/BdxTransferServer.cpp

```diff
@@ -0,0 +1,209 @@
+}
+
+void BDXTransferServer::Shutdown()
+{
+    VerifyOrReturn(nullptr != mSystemLayer);
+    VerifyOrReturn(nullptr != mExchangeMgr);
+
+    LogErrorOnFailure(mExchangeMgr->UnregisterUnsolicitedMessageHandlerForType(MessageType::SendInit));
+
+    mSystemLayer = nullptr;
+    mExchangeMgr = nullptr;
+}
+
+void BDXTransferServer::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    ChipLogDetail(BDX, "Got an event %s", event.ToString(event.EventType));
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kInitReceived:
+        AbortTransferOnFailure(OnTransferSessionBegin(event));
+        break;
+    case TransferSession::OutputEventType::kStatusReceived:
+        ChipLogError(BDX, "Got StatusReport %x", static_cast<uint16_t>(event.statusData.statusCode));
+        LogErrorOnFailure(OnTransferSessionEnd(CHIP_ERROR_INTERNAL));
+        break;
+    case TransferSession::OutputEventType::kInternalError:
+        LogErrorOnFailure(OnTransferSessionEnd(CHIP_ERROR_INTERNAL));
+        break;
+    case TransferSession::OutputEventType::kTransferTimeout:
+        LogErrorOnFailure(OnTransferSessionEnd(CHIP_ERROR_TIMEOUT));
+        break;
+    case TransferSession::OutputEventType::kBlockReceived:
+        AbortTransferOnFailure(OnBlockReceived(event));
+        break;
+    case TransferSession::OutputEventType::kMsgToSend:
+        LogErrorOnFailure(OnMessageToSend(event));
+
+        if (event.msgTypeData.HasMessageType(MessageType::BlockAckEOF))
+        {
+            LogErrorOnFailure(OnTransferSessionEnd(CHIP_NO_ERROR));
+        }
+        break;
+    case TransferSession::OutputEventType::kAckEOFReceived:
+    case TransferSession::OutputEventType::kNone:
+    case TransferSession::OutputEventType::kQueryWithSkipReceived:
+    case TransferSession::OutputEventType::kQueryReceived:
+    case TransferSession::OutputEventType::kAckReceived:
+    case TransferSession::OutputEventType::kAcceptReceived:
+        // Nothing to do.
+        break;
+    default:
+        // Should never happen.
+        chipDie();
+        break;
+    }
+}
+
+CHIP_ERROR BDXTransferServer::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                System::PacketBufferHandle && payload)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(ec != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // If we receive a SendInit message, then we prepare for transfer
+    if (payloadHeader.HasMessageType(MessageType::SendInit))
+    {
+        FabricIndex fabricIndex = ec->GetSessionHandle()->GetFabricIndex();
+        NodeId peerNodeId       = ec->GetSessionHandle()->GetPeer().GetNodeId();
+        VerifyOrReturnError(fabricIndex != kUndefinedFabricIndex, CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(fabricIndex != kUndefinedFabricIndex, CHIP_ERROR_INVALID_ARGUMENT);
```

Comment:
Shouldn't this second check be for the node id?  We already checked the fabric index.

Reply:
Sigh. Thanks.